File: 20.valid-parentheses.go
package main

import "fmt"

/*
 * @lc app=leetcode id=20 lang=golang
 *
 * [20] Valid Parentheses
 */

// @lc code=start
func isValid(s string) bool {
	var a, b, c int
	for _, v := range s {
		switch v {
		case '(':
			a++
		case '[':
			b++
		case '{':
			c++
		case ')':
			a--
		case ']':
			b--
		case '}':
			c--
		}
		if a < 0 || b < 0 || c < 0 {
			return false
		}
	}
	return a == 0 && b == 0 && c == 0
}

// @lc code=end

func main() {
	fmt.Println(isValid("([])"))
}

File: 20.valid-parentheses.cpp
/*
 * @lc app=leetcode id=20 lang=cpp
 *
 * [20] Valid Parentheses
 */

#include "leetcode.h"

// @lc code=start
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> map;
        map[')'] = '(';
        map['}'] = '{';
        map[']'] = '[';

        stack<char> tmp;
        for (int i = 0; i < s.size(); i++) {
            if (s[i] == '(' || s[i] == '{' || s[i] == '[') 
                tmp.push(s[i]);
            else if (!tmp.empty() && tmp.top() == map[s[i]]) 
                tmp.pop();
            else 
                return false;
        }

        return tmp.size() == 0;
    }
};
// @lc code=end

